### PR TITLE
Midi signals

### PR DESCRIPTION
**Pull request by rath-isaac**

### CODE REVIEW

File: semantic.py [hunk 1]
```diff
@@ -381,16 +394,143 @@ def is_valid_program(self, tree):
...
+            return False
+
+        signals = []
+        signals.append({'type':'measure'})
+        
+        for i in tree.children:
+            if i.data == 'instrumentation':
+                signals += (self.instrumentation_to_signal(i))
+        
+        print("\n\n\n\n\n\n\n")
```
Comment (mpfeifer1) [inline semantic.py]: ummmmm no

File: semantic.py [hunk 2]
```diff
@@ -381,16 +394,143 @@ def is_valid_program(self, tree):
 
     # Takes in a measure, builds a list of signals
     def measure_to_signal(self, tree):
-        pass
+        if tree.data != 'measure':
```
Comment (mpfeifer1) [inline semantic.py]: call isvalidmeasure

File: semantic.py [hunk 3]
```diff
@@ -381,16 +394,143 @@ def is_valid_program(self, tree):
...
+                signals += (self.instrumentation_to_signal(i))
+        
+        print("\n\n\n\n\n\n\n")
+        for x in signals:
+            print(x)
+        
+        return signals
+
+    def instrumentation_to_signal(self, tree):
+        if tree.data != 'instrumentation':
```
Comment (mpfeifer1) [inline semantic.py]: call isvalidinstrumentation

File: semantic.py [hunk 4]
```diff
@@ -381,16 +394,143 @@ def is_valid_program(self, tree):
...
+            print('error: not an instrumentation')
+            return False
 
+        signals = []
+        name = tree.children[0]
+        
+        signals.append({'type':'instrument', 'name':str(name)})
+        
+        #print('\ninstrumentation: ' + tree.children[0])
```
Comment (mpfeifer1) [inline semantic.py]: can probably remove this guy, too

File: semantic.py [hunk 5]
```diff
@@ -381,16 +394,143 @@ def is_valid_program(self, tree):
...
+        if instrumentToNumber.__contains__(tree.children[0]):
```
Comment (mpfeifer1) [inline semantic.py]: i think you can say
if x in y:
    pass
rather than using contains

File: semantic.py [hunk 6]
```diff
@@ -381,16 +394,143 @@ def is_valid_program(self, tree):
...
+            for i in tree.children[1:]:
+                signals += (self.noteitem_to_signal(i))
+        else:
+            print('invalid instrument')
+
+        return signals
+
+    def noteitem_to_signal(self,tree):
+        if tree.data != 'noteitem':
```
Comment (mpfeifer1) [inline semantic.py]: isvalidnoteitem

File: semantic.py [hunk 7]
```diff
@@ -381,16 +394,143 @@ def is_valid_program(self, tree):
...
+            print('error: invalid noteitem')
+        
+        signals = []
+
+        for i in tree.children:
+            # possible noteitem children : note , inlinedynamic
+            if i.data == 'note':
+                signals += (self.note_to_signal(i))
```
Comment (mpfeifer1) [inline semantic.py]: drop the extra parenthesis on these guys

File: semantic.py [hunk 8]
```diff
@@ -381,16 +394,143 @@ def is_valid_program(self, tree):
...
+            elif i.data == 'inlinedynamic':
+                signals += (self.inlinedynmaic_to_signal(i))
+            else:
+                print('invalid noteitem child')
+   
+        return signals
+ 
+    def note_to_signal(self, tree):
+        if tree.data != 'note':
```
Comment (mpfeifer1) [inline semantic.py]: isvalidnote ;)

File: semantic.py [hunk 9]
```diff
@@ -381,16 +394,143 @@ def is_valid_program(self, tree):
...
+            print('error: not a note!')
+
+        signals = []
```
Comment (mpfeifer1) [inline semantic.py]: just add a quick comment explaining what this is doing

File: semantic.py [hunk 10]
```diff
@@ -381,16 +394,143 @@ def is_valid_program(self, tree):
...
+                chordsig['length_denom'] = int(den)
+                signals.append(chordsig)
+            elif i.data == "tuple":
+                signals += self.tuple_to_signal(i)
+            else:
+                print("invalid note child")
+
+        return signals
+
+    def notename_to_signal(self, tree):
```
Comment (mpfeifer1) [inline semantic.py]: isvalidnotename

File: semantic.py [hunk 11]
```diff
@@ -381,16 +394,143 @@ def is_valid_program(self, tree):
...
+            if 'Token' in str(type(i)):
+                name += str(i)
+            elif 'Tree' in str(type(i)): #it's a token
+                name+=i.children[0]
+            else:
+                print('invalid notename child')
+        
+        return name
+
+    def inlinedynmaic_to_signal(self, tree):
```
Comment (mpfeifer1) [inline semantic.py]: isvalid...

File: semantic.py [hunk 12]
```diff
@@ -381,16 +394,143 @@ def is_valid_program(self, tree):
...
+        return [{'type':'dynamic', 'volume':str(tree.children[0])}]
+        
     def chord_to_signal(self, tree):
-        pass
+        if tree.data != 'chord':
```
Comment (mpfeifer1) [inline semantic.py]: isvalid

File: semantic.py [hunk 13]
```diff
@@ -381,16 +394,143 @@ def is_valid_program(self, tree):
...
+            print('error! not a chord')
+        
+        notes = []
+
+        for i in tree.children:
+            # only children of a chord are notenames
```
Comment (mpfeifer1) [inline semantic.py]: isvalid

File: semantic.py [hunk 14]
```diff
@@ -381,16 +394,143 @@ def is_valid_program(self, tree):
...
+            if i.data == 'notename':
+                notes.append(self.notename_to_signal(i))
+            else:
+                print('invalid chord child')
+        
+        return notes
 
     def tuple_to_signal(self, tree):
-        pass
+        if tree.data != 'tuple':
```
Comment (mpfeifer1) [inline semantic.py]: you know the drill by this point

File: semantic.py [hunk 15]
```diff
@@ -381,16 +394,143 @@ def is_valid_program(self, tree):
...
+            print('error: not a tuple')
+        # put dummy data in a tuple signal because we don't like them much
```
Comment (evan-burandt) [inline semantic.py]: This should probably be actual data in the future, but for now it is fine.

File: semantic.py (hunk 2 above)
Comment (rath-isaac) [reply semantic.py]: None is the isvalid stuff is in this branch and will be sorted later

**Pull request merged**